File: gke_customization/gke_hrms/report/pf_challan_report/pf_challan_report.js
// Copyright (c) 2025, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["PF Challan Report"] = {
	"filters": [
		{
            "label": __("Month"),
            "fieldtype": "Select",
            "fieldname": "month",
            "reqd": 1,
            "options": [
                "January","February","March","April",
                "May","June","July","August",
                "September","October","November","December"
            ],
            "default": new Date().toLocaleString('en-US', { month: 'long' }),
            "on_change": function() {
                set_dates();
            }
        },
        {
            "label": __("Year"),
            "fieldtype": "Select",
            "fieldname": "year",
            "reqd": 1,
            "options": (() => {
                let years = [];
                let currentYear = new Date().getFullYear();
                for (let i = currentYear - 5; i <= currentYear + 5; i++) {
                    years.push(String(i));
                }
                return years;
            })(),
            "default": String(new Date().getFullYear()),
            "on_change": function() {
                set_dates();
            }
        },
		{
			"label": __("From Date"),
			"fieldtype": "Date",
			"fieldname": "from_date",
			"read_only": 1,
			"default": frappe.datetime.month_start(frappe.datetime.get_today()),
			"on_change": function(query_report) {
				var from_date = query_report.get_values().from_date;
				if (!from_date) {
					return;
				}
				let date = new moment(from_date)
				var to_date = date.endOf('month').format()
				query_report.set_filter_value({
					"to_date": to_date
				});
			}
		},
		{
			"label": __("To Date"),
			"fieldtype": "Date",
			"fieldname": "to_date",
			"reqd": 1,
			"read_only": 1,
			"default": frappe.datetime.month_end(frappe.datetime.get_today())
		},
		{
			"label": __("Company"),
			"fieldtype": "Link",
			"fieldname": "company",
			"options": "Company",
			"reqd": 1,
		},
		{
			"label": __("Branch"),
			"fieldtype": "Link",
			"fieldname": "branch",
			"options": "Branch" 
		}, 
	],
	onload: (report) => {
		set_dates();
		report.page.add_button("Clear Filters", function() {
			window.open("/app/query-report/PF%20Challan%20Report", "_self")
		}).addClass("btn-info")	
		report.page.add_inner_button(__("PF File"), function () {
			frappe.call({
				method: "gke_customization.gke_hrms.report.pf_challan_report.pf_challan_report.export_txt", 
				args: {
					filters: report.get_values()
				},
				callback: function (r) {
				if (r.message) {
					// 🔽 Auto download
					const link = document.createElement("a");
					link.href = r.message;
					link.download = "pf_export.txt";
					document.body.appendChild(link);
					link.click();
					document.body.removeChild(link);
				}
			}
			});
		}).addClass("btn-secondary");
		report.page.add_inner_button(__("PF Contribution Report"), function () {
			var filters = report.get_values();
			frappe.set_route("query-report", "PF Contribution Report", { month: filters.month, year: filters.year, company: filters.company });
		}); 
	},
	"after_datatable_render": function (datatable_obj) {
		const report = frappe.query_report;
		ensure_panel_inserted(report);
		fetch_and_render_account_total_panel(report);
	},
 
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		return value;
	},
};
 
function ensure_panel_inserted(report) {
	if (report.pf_panel && report.pf_panel.length && report.pf_panel.closest("body").length) {
		return;
	}
 
	const $panel = $('<div class="pf-account-total-panel"></div>');
 
	let $mount = report.page.main.find(".report-wrapper");
	if (!$mount.length) {
		$mount = report.page.main.find(".dt-scrollable, .datatable");
	}
	if (!$mount.length) {
		$mount = report.page.main;
	}
 
	$mount.first().prepend($panel);
	report.pf_panel = $panel;
}
 
function fetch_and_render_account_total_panel(report) {
	const filters = report.get_values();
	if (!filters || !filters.company) return;
 
	frappe.call({
		method: "gke_customization.gke_hrms.report.pf_challan_report.pf_challan_report.get_account_total_summary",
		args: { filters: filters },
		callback: function (r) {
			if (r.message) {
				render_account_total_panel(report, r.message, filters);
			}
		},
	});
}
 
function render_account_total_panel(report, summary, filters) {
	const fmt = (n) => {
		const num = Number(n) || 0;
		return num.toLocaleString("en-IN");
	};
 
	const ac1 = summary.ac1 || {};
	const ac10 = summary.ac10 || {};
	const ac21 = summary.ac21 || {};
	const c = summary.contribution || {};
 
	const month = filters.month || ""; 
	const html = `
	<div style="padding: 1rem 0;">
		<div style="background: var(--bg-color); border-radius: var(--border-radius-lg); border: 0.5px solid var(--border-color); overflow: hidden;">
 
			<div style="padding: 0.875rem 1.25rem; border-bottom: 0.5px solid var(--border-color); display: flex; align-items: center; justify-content: space-between;">
				<span style="font-size: 16px; font-weight: 500;">Account total</span>
				<span style="font-size: 13px; color: var(--text-muted);">${frappe.utils.escape_html(month)}</span>
			</div>
 
			<div style="display: grid; grid-template-columns: repeat(3, minmax(0, 1fr)); border-bottom: 0.5px solid var(--border-color);">
 
				<div style="padding: 1rem 1.25rem; border-right: 0.5px solid var(--border-color);">
					<div style="display: inline-block; font-size: 13px; font-weight: 500; padding: 4px 14px; border-radius: 999px; background: #17a2b8; color: var(--primary); margin-bottom: 14px;">A/C 1</div>
					<div style="display: flex; align-items: baseline; gap: 10px; margin-bottom: 6px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Employee</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac1.subscribers)}</span>
					</div>
					<div style="display: flex; align-items: baseline; gap: 10px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Total Wages</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac1.wages)}</span>
					</div>
				</div>
 
				<div style="padding: 1rem 1.25rem; border-right: 0.5px solid var(--border-color);">
					<div style="display: inline-block; font-size: 13px; font-weight: 500; padding: 4px 14px; border-radius: 999px; background: #17a2b8; color: var(--primary); margin-bottom: 14px;">A/C 10</div>
					<div style="display: flex; align-items: baseline; gap: 10px; margin-bottom: 6px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Employee</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac10.subscribers)}</span>
					</div>
					<div style="display: flex; align-items: baseline; gap: 10px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Total Wages</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac10.wages)}</span>
					</div>
				</div>
 
				<div style="padding: 1rem 1.25rem;">
					<div style="display: inline-block; font-size: 13px; font-weight: 500; padding: 4px 14px; border-radius: 999px; background: #17a2b8; color: var(--primary); margin-bottom: 14px;">A/C 21</div>
					<div style="display: flex; align-items: baseline; gap: 10px; margin-bottom: 6px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Employee</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac21.subscribers)}</span>
					</div>
					<div style="display: flex; align-items: baseline; gap: 10px;">
						<span style="font-size: 13px; color: var(--text-muted); min-width: 78px;">Total Wages</span>
						<span style="font-size: 17px; font-weight: 500;">${fmt(ac21.wages)}</span>
					</div>
				</div>
 
			</div>
 
			<div style="display: grid; grid-template-columns: repeat(7, minmax(0, 1fr)); border-bottom: 0.5px solid var(--border-color);">
 
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px; line-height: 1.3;">A/C No.1 Employee</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_1_employee)}</p>
				</div>
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px; line-height: 1.3;">A/C No.1 Employer</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_1_employer)}</p>
				</div>
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px;">A/C No.2</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_2)}</p>
				</div>
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px;">A/C No.10</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_10)}</p>
				</div>
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px;">A/C No.21</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_21)}</p>
				</div>
				<div style="padding: 0.875rem 1rem; border-right: 0.5px solid var(--border-color);">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px;">A/C No.22</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(c.ac_no_22)}</p>
				</div>
				<div style="padding: 0.875rem 1rem;">
					<p style="font-size: 12px; color: var(--text-muted); margin: 0 0 4px;">Total</p>
					<p style="font-size: 15px; font-weight: 500; margin: 0;">${fmt(summary.grand_total)}</p>
				</div>
			</div>
 
			<div style="padding: 0.875rem 1.25rem; display: flex; align-items: center; gap: 28px; flex-wrap: wrap;">
				<div style="display: flex; align-items: baseline; gap: 8px;">
					<span style="font-size: 15px; font-weight: 500; color: var(--text-muted);">Total employees</span>
					<span style="font-size: 14px;">${fmt(summary.employee_count)}</span>
				</div>
				<div style="display: flex; align-items: baseline; gap: 8px;">
					<span style="font-size: 15px; font-weight: 500; color: var(--text-muted);">Gross pay</span>
					<span style="font-size: 14px;">${fmt(summary.gross_pay)}</span>
				</div>
				<div style="display: flex; align-items: baseline; gap: 8px;">
					<span style="font-size: 15px; font-weight: 500; color: var(--text-muted);">PF amount</span>
					<span style="font-size: 14px;">${fmt(summary.pf_amount)}</span>
				</div>
				
				
			</div>
 
		</div>
	</div>
	`;
	// <div style="display: flex; align-items: baseline; gap: 8px; margin-left: auto;">
	// 				<span style="font-size: 14px; font-weight: 500;">Total</span>
	// 				<span style="font-size: 19px; font-weight: 500;">${fmt(summary.grand_total)}</span>
	// 			</div>
	
	if (report.pf_panel) {
		report.pf_panel.html(html);
	}
}
 

function fetch_month_list() {
	frappe.call({
		method: "gurukrupa_customizations.gurukrupa_customizations.report.monthly_in_out.monthly_in_out.get_month_range",
		freeze: false,
		callback: function(r) {
			var month = frappe.query_report.get_filter('month');
				month.df.options.push(...r.message);
				month.refresh();
		}
	})
}

function set_dates() {
    let report = frappe.query_report;
    if (!report) return;

    let monthName = report.get_filter_value("month");
    let year = parseInt(report.get_filter_value("year"));

    if (!monthName || !year) return;

    const monthIndex = [
        "January","February","March","April",
        "May","June","July","August",
        "September","October","November","December"
    ].indexOf(monthName);

    let from_date = new Date(year, monthIndex, 1);
    let to_date = new Date(year, monthIndex + 1, 0);

    report.set_filter_value({
        from_date: frappe.datetime.obj_to_str(from_date),
        to_date: frappe.datetime.obj_to_str(to_date)
    });
}

File: gke_customization/gke_hrms/report/pf_challan_report/pf_challan_report.py
# Copyright (c) 2025, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import get_site_path,today, getdate
import os

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_data(filters=None):
	from_date = filters.get("from_date")
	to_date = filters.get("to_date")
	company = filters.get("company")
	branch = filters.get("branch")

	conditions = []
	values = {
		"from_date": from_date,
		"to_date": to_date,
		"company": company
	} 
    
	if branch:
		conditions.append("AND e.branch = %(branch)s")
		values["branch"] = branch
    
	condition_query = ""
	if conditions:
		condition_query = " AND " + " AND ".join(conditions)

	# frappe.throw(f"{values}")
	query = f"""
		SELECT 
			ss.company AS company, 
			ss.branch AS branch, 
			ss.department AS department, 
			ss.designation AS designation,  
			ss.employee AS employee_id, 
			ss.employee_name AS employee_name,
            ss.custom_gross_salary,
			ss.absent_days,
			e.date_of_birth,
			ss.payment_days,
			ss.total_working_days,
			/* Gross Pay counted once */
			MAX(ss.gross_pay) AS gross_pay,

			/* PF amount only if age < 58 */
			SUM(
				CASE 
					WHEN sd.salary_component = 'Provident Fund'
						#  AND TIMESTAMPDIFF(YEAR, e.date_of_birth, ss.end_date) < 58
					THEN sd.amount
					ELSE 0
				END
			) AS pf_amount,

			MAX(
			    CASE 
			        WHEN sd.salary_component = 'Basic' 
					THEN sd.amount
					ELSE 0
				END
			) AS epf_wages,
			
			MAX(
			    CASE 
			        WHEN sd.salary_component = 'Advance' 
					THEN sd.amount
					ELSE 0
				END
			) AS refund_advance,

			e.gender AS gender

		FROM 
			`tabSalary Slip` ss
		JOIN 
			`tabSalary Detail` sd 
				ON ss.name = sd.parent
		JOIN 
			`tabEmployee` e 
				ON ss.employee = e.name

		WHERE 
			ss.company = %(company)s
			AND ss.start_date = %(from_date)s
			AND ss.end_date = %(to_date)s
			# AND ss.docstatus = 1
			# AND ss.status = 'Submitted'

			/* Only employees whose PF is deducted */
			AND EXISTS (
				SELECT 1
				FROM `tabSalary Detail` sd_pf
				WHERE sd_pf.parent = ss.name
				  AND sd_pf.salary_component = 'Provident Fund'
				  AND sd_pf.amount > 0
			)
			{condition_query}
		GROUP BY ss.employee
	"""

	data = frappe.db.sql(query, values, as_dict=True)
	totals = get_totals(data, filters.get("employee"))
	data += totals
	
	return data


def get_totals(data, employee=None):
	totals = {
		"company": "Total",
		"branch": "",
		"employee_id": "",
		"employee_name": "Employee Count: {}".format(len(data)),
		"employee_count": len(data),
		"gross_pay": sum((row.get("gross_pay") or 0) for row in data),
		"pf_amount": sum((row.get("pf_amount") or 0) for row in data),
	}
	return [totals]

def get_columns(filters=None):
	columns = [
		{
			"label": "Company",
			"fieldname": "company",
			"fieldtype": "Data",
			"width": 200
		}, 
		{
			"label": "Branch",
			"fieldname": "branch",
			"fieldtype": "Data",
			"width": 200
		}, 
		{
			"label": "Employee",
			"fieldname": "employee_id",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": "Employee Name",
			"fieldname": "employee_name",
			"fieldtype": "Data",
			"width": 200
		},
		{
			"label": "Gross Pay",
			"fieldname": "gross_pay",
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"label": "PF Amount",
			"fieldname": "pf_amount",
			"fieldtype": "Currency",
			"width": 120
		}
		 
	]
	return columns

@frappe.whitelist()
def export_txt(filters=None):
	if isinstance(filters, str):
		filters = frappe.parse_json(filters)

	data = get_data(filters)

	lines = []
	for row in data:
	# skip total row if present
		if row.get("company") == "Total":
			continue

		age = 0
		dob = row.get("date_of_birth")
		if dob:
			today_date = getdate(today())
			dob_date = getdate(dob)
			age = today_date.year - dob_date.year - (
				(today_date.month, today_date.day) < (dob_date.month, dob_date.day)
			)

		# only include employees with age < 58
		if row.get("employee_id") :
			# and age < 58:
			gross_pay = round(row.get("gross_pay") or 0)
			uan_number = frappe.db.get_value("Employee", row["employee_id"], "uan_number")
			name_as_per_aadhar = frappe.db.get_value("Employee", row["employee_id"], "name_as_per_aadhar")
			# epf_wages = round(row.get("epf_wages") or 0)
			
			basic_ear = round(row.get("epf_wages") or 0)
			eps_wages = 0
			
			epf_wages = 15000 if basic_ear > 15000 else basic_ear

			if age >= 58:
				eps_wages = 0
			else:
				eps_wages = 15000 if basic_ear > 15000 else basic_ear

			edli_wages = 15000 if basic_ear > 15000 else basic_ear
			ctc = round(float(row.get("custom_gross_salary") or 0)) 

			if basic_ear < 15000:
				epf_contribution = round((basic_ear or 0) * 0.12)
				if age >= 58:
					eps_contribution = 0
				else:
					eps_contribution = round((basic_ear or 0) * 0.0833)
			else:	
				epf_contribution = round(row.get("pf_amount") or 0)
				eps_contribution = round(15000 * 0.0833)
			contribution = (epf_contribution - eps_contribution) or 0


			ncp = 0
			total_working_days = row.get("total_working_days") or 0
			payment_days = row.get("payment_days") or 0

			if row.get("absent_days"):
				ncp = round(row.get("absent_days") or 0)
			else:
				ncp = round(total_working_days - payment_days)

			refund_advance = row.get("refund_advance") or 0

			# 100982928420#~#Bhikhubhai Maheshbhai Modi#~#22589#~#14065
			# #~#14065#~#14065#~#1688
			# #~#1688#~#0#~#0#~#0

			lines.append(
				f"{uan_number}#~#{name_as_per_aadhar}#~#{gross_pay}#~#{epf_wages}"
				f"#~#{eps_wages}#~#{edli_wages}#~#{epf_contribution}"
				f"#~#{eps_contribution}#~#{contribution}#~#{ncp}#~#{refund_advance}"
			)

	content = "\n".join(lines)

	file_name = "pf_export.txt"
	file_path = get_site_path("public", "files", file_name)

	with open(file_path, "w", encoding="utf-8") as f:
		f.write(content)

	return f"/files/{file_name}"

# def export_txt(filters=None):
# 	if isinstance(filters, str):
# 		filters = frappe.parse_json(filters)

# 	data = get_data(filters)

# 	lines = []
# 	for row in data:
# 		if row.get("company") == "Total":
#             pass

# 		age = 0
# 		dob = row.get("date_of_birth")
# 		if dob:
# 			today_date = getdate(today())
# 			dob_date = getdate(dob)
# 			age = today_date.year - dob_date.year - (
# 				(today_date.month, today_date.day) < (dob_date.month, dob_date.day)
# 			)
# 		if row.get("employee_id") and age < 58:
# 			gross_pay = round(row.get("gross_pay"))
# 			uan_number = frappe.db.get_value("Employee", row["employee_id"], "uan_number")
# 			name_as_per_aadhar = frappe.db.get_value("Employee", row["employee_id"], "name_as_per_aadhar")
# 			epf_wages = round(row['epf_wages']) or 0
# 			eps_wages = epf_wages
# 			edli_wages = epf_wages
# 			epf_contribution = round((epf_wages or 0) * 0.12)
# 			eps_contribution = round((epf_wages or 0) * 0.0833)
# 			contribution = (epf_contribution - eps_contribution) or 0
# 			ncp = row.get("absent_days") or 0
# 			refund_advance = row.get("refund_advance") or 0

		
# 		lines.append(
# 			f"{uan_number}#~#{name_as_per_aadhar}#~#{gross_pay}#~#{epf_wages}#~#{eps_wages}#~#{edli_wages}#~#{epf_contribution}#~#{eps_contribution}#~#{contribution}#~#{ncp}#~#{refund_advance}"
# 		)

# 	content = "\n".join(lines)

# 	file_name = "pf_export.txt"
# 	file_path = get_site_path("public", "files", file_name)

# 	with open(file_path, "w", encoding="utf-8") as f:
# 		f.write(content)

# 	return f"/files/{file_name}"